### PR TITLE
[kernel] Support new KCache Layout - Context Attention Triton Kernel

### PR DESCRIPTION
**Pull request by yuanheng-zhao**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

New Cache Layout is proposed in this PR https://github.com/hpcaitech/ColossalAI/pull/5643, which improves CUDA Flash Decoding performance.

To modify the whole flows with revised kcache layout (Triton kernel flow and CUDA kernel flows), we're gonna to revise kcache-relatable kernels.

In this PR, support for the new KCache layout is added in context attention triton kernel. The alibi kernel is not updated yet as it's pretty similar to the regular context attention kernel, and it will be refactored and updated as the whole triton flow for the new KCache layout has been supported to reduce redundancy.   

pytest_infer_0

pytest_infer_1

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/kernel/triton/context_attn_unpad.py [hunk 1]
```diff
@@ -375,21 +553,33 @@ def context_attention_unpadded(
     q: torch.Tensor,  # [num_tokens, num_heads, head_dim]
     k: torch.Tensor,  # [num_tokens, num_kv_heads, head_dim]
     v: torch.Tensor,  # [num_tokens, num_kv_heads, head_dim]
-    k_cache: torch.Tensor,  # [num_blocks, num_kv_heads, head_dim, block_size]
-    v_cache: torch.Tensor,  # [num_blocks, num_kv_heads, head_dim, block_size]
+    k_cache: torch.Tensor,  # [num_blocks, num_kv_heads, block_size, head_dim]
```
Comment (yuehuayingxueluo) [inline colossalai/kernel/triton/context_attn_unpad.py:556]: Is there a problem with the shape here?

File: colossalai/kernel/triton/context_attn_unpad.py [hunk 2]
```diff
@@ -185,6 +185,184 @@ def _fwd_context_paged_attention_kernel(
...
+# merge `_fwd_context_paged_attention_kernel_v2` with `_fwd_context_paged_attention_kernel` later
+# as the kcache layout has been supported in the whole triton flow.
+@triton.jit
+def _fwd_context_paged_attention_kernel_v2(
+    Q,
+    K,
+    V,
+    O,
+    KCache,  # [num_blocks, num_kv_heads, block_size, head_dim]
+    VCache,  # [num_blocks, num_kv_heads, head_dim // x, block_size, x]
```
Comment (yuehuayingxueluo) [inline colossalai/kernel/triton/context_attn_unpad.py:199]: The shapes of the key cache and value cache should be swapped.

Comment (yuanheng-zhao) [reply colossalai/kernel/triton/context_attn_unpad.py]: Revised

**Review by yuehuayingxueluo — APPROVED** [review]
LGTM

**Review by SunflowerAries — APPROVED** [review]
LGTM

**Pull request merged**